### DISCUSSION TITLE
How to pass through values from defaultValues to handleSubmit without rendering fields

### DISCUSSION THREAD

**Discussion by delijah** [react-hook-form/react-hook-form#6240] [Q&A]
Hello

We are currently upgrading from 6.15.0 to 7.12.2 and are experiencing troubles with `shouldUnregister: false`. We need to set it to false because of how some forms work in our stack.

I'll try to explain: Basically we pass on all the data we get from an API to our forms. Only some specific parts of the data is really mutable for the user by using the form. Therefore we provide fields to the user, so he can change those parts of the data. Our API expects the whole payload back, including the modified and non-modified data.

As already explained.. so far we just passed all the data to the form, defining some form fields for the user to edit the data, and the result we got was the updated data we could send back to the API right away.

Now we've realized, that there were some major changes in v7 regarding this. It looks like v7 registers every key of the defaultValues data object as a field, even though we do NOT render a field for that part of the data object (even for nested data structures!!).

We wonder if there is an option to disable this "magic behind the scenes"? I think the best would be if react-hook-form would not mess around with data it does not know anything about it (if there is nothing explicitly defined by the developer).

Please help :)

I know there are a lot discussions going on about this or similar thingies, but didn't find the right one to join in..

**Comment by bluebill1049** [comment]
try to set `shouldUnregister: true`

**Reply by irynakoshynska** [reply]
thank you :) yes, it helps 
but we have conditionally rendered fields, and in order not to loose data, we need to toggle the visibility, right?

**Reply by bluebill1049** [reply]
That's right.

**Reply by delijah** [reply]
hmmmm... ok, cool this would help us already, thx :)

...but somehow this also sounds a bit like a workaround to me. the option `shouldUnregister: false` implies, that fields are not unregistered, but it does not really say that some magic field registering will happen in the background. of course now i know, because i dug through several discussions and we've talked together. but i think it might be more intuitive if we would have a second option, something like `autoRegister: true` to make the magic work.

so we could use for example:
```
shouldUnregister: false,
autoRegister: false,
```

what do you think about it?

**Reply by delijah** [reply]
We've also just realized, we are having lots of fields that have some business logic behind them. So if the react component is rendered, but turned invisible with css (following your proposal), it would still execute that logic. For example validation for required fields, or fetching data from the API for dropdown lists... so the job is not really done with just turning those fields invisible...

**Reply by bluebill1049** [reply]
why not use default `shouldUnregister` then? and `unregister` input when you want to remove that input.

**Reply by delijah** [reply]
Hmmm... so you are saying we should use `shouldUnregister: false` (which is the default value) and execute unregister for every key in our data object which is not editable and therefore we are not rendering any form field (in order to tell the form to not use it as a form field to avoid unwanted data modification by react-hook-form)?

**Reply by delijah** [reply]
At the moment we are stuck with the following solutions, but none of them are really appealing to us since all of them would mean a lot of work and it would add unnecessary complexity to our stack. Still, not having this register magic built inside react-hook-form looks like the cleanest solution to us.

- use shouldUnregister: true; refactor components to turn invisible and for not doing anything (unregister to avoid validation, do no api requests) when invisible
- use shouldUnregister: false; only pass editable data to forms
- use shouldUnregister: false; create a hidden field for all the non modifiable data
- react-hook-form changes it’s ways
-- we fork react-hook-form and remove that registering of data

If anybody sees a better solution we highly appreciate your help.

**Reply by delijah** [reply]
@bluebill1049 i've created a PR as an example: https://github.com/react-hook-form/react-hook-form/pull/6271
